src/middlewared/middlewared/api/v25_10_0/pool_snapshot.py

```diff
@@ -0,0 +1,178 @@
+    "PoolSnapshotCreateResult", "PoolSnapshotDeleteArgs", "PoolSnapshotDeleteResult", "PoolSnapshotHoldArgs",
+    "PoolSnapshotHoldResult", "PoolSnapshotReleaseArgs", "PoolSnapshotReleaseResult", "PoolSnapshotRollbackArgs",
+    "PoolSnapshotRollbackResult", "PoolSnapshotUpdateArgs", "PoolSnapshotUpdateResult",
+]
+
+
+def validate_snapshot_naming_schema_and_return(value: NonEmptyString):
+    validate_snapshot_naming_schema(value)
+    return value
+
+
+ReplicationSnapshotNamingSchema = Annotated[NonEmptyString, AfterValidator(validate_snapshot_naming_schema_and_return)]
+UserPropertyKey = Annotated[str, Field(pattern=r'.*:.*')]
+
+
+class PoolSnapshotEntryPropertyFields(BaseModel):
+    value: str
+    rawvalue: str
+    source: Literal["NONE", "DEFAULT", "LOCAL", "TEMPORARY", "INHERITED", "RECEIVED"]
+    parsed: Any
+
+
+class PoolSnapshotHoldTag(BaseModel):
+    truenas: int = NotRequired
+    """Present if a hold has been placed on the snapshot."""
+
+
+class PoolSnapshotEntry(BaseModel):
+    properties: dict[str, PoolSnapshotEntryPropertyFields]
+    pool: str
+    name: str
+    type: Literal["SNAPSHOT"]
+    snapshot_name: str
+    dataset: str
+    id: str
+    createtxg: str
+    holds: PoolSnapshotHoldTag = NotRequired
+    """Returned when options.extra.holds is set."""
+
+
+class PoolSnapshotCreateUpdateEntry(PoolSnapshotEntry):
+    holds: Excluded = excluded_field()
+
+
+class PoolSnapshotCreateTemplate(BaseModel, ABC):
+    dataset: NonEmptyString
+    recursive: bool = False
+    exclude: list[str] = []
+    vmware_sync: bool = False
+    properties: dict = {}
+
+
+class PoolSnapshotCreateWithName(PoolSnapshotCreateTemplate):
+    name: NonEmptyString
+
+
+class PoolSnapshotCreateWithSchema(PoolSnapshotCreateTemplate):
+    naming_schema: ReplicationSnapshotNamingSchema
+
+
+class PoolSnapshotDeleteOptions(BaseModel):
+    defer: bool = False
+    """Defer deletion of the snapshot."""
+    recursive: bool = False
+
+
+class PoolSnapshotHoldOptions(BaseModel):
+    recursive: bool = False
+    """Hold snapshots recursively."""
+
+
+class PoolSnapshotReleaseOptions(BaseModel):
+    recursive: bool = False
+    """Release snapshots recursively. Only the tags that are present on the parent snapshot will be removed."""
+
+
+class PoolSnapshotRollbackOptions(BaseModel):
+    recursive: bool = False
+    """Destroy any snapshots and bookmarks more recent than the one specified."""
+    recursive_clones: bool = False
+    """Just like `recursive`, but also destroy any clones."""
+    force: bool = False
+    """Force unmount of any clones."""
+    recursive_rollback: bool = False
+    """Do a complete recursive rollback of each child snapshot for `id`. If any child does not have specified snapshot,
+    this operation will fail."""
+
+
+class PoolSnapshotUserPropertyUpdate(BaseModel):
+    key: UserPropertyKey
+    value: str
+
+
+class PoolSnapshotUpdate(BaseModel):
```

Comment:
@yocalebo Should we allow snapshot renames via the update API? This would resolve one complaint / feature request related to renames.